skgstat/MetricSpace.py

```diff
@@ -243,3 +255,107 @@ def dists(self):
 
         # return
         return self._dists
+
+class ProbabalisticMetricSpace(MetricSpace):
+    """Like MetricSpace but samples the distance pairs only returning a
+       `samples` sized subset. `samples` can either be a fraction of
+       the total number of pairs (float < 1), or an integer count.
+    """
+    
+    def __init__(self, coords, dist_metric="euclidean", max_dist=None, samples=0.5):
+        """ProbabalisticMetricSpace class
+
+        Parameters
+        ----------
+        coords : numpy.ndarray
+            Coordinate array of shape (Npoints, Ndim)
+        dist_metric : str
+            Distance metric names as used by scipy.spatial.distance.pdist
+        max_dist : float
+            Maximum distance between points after which the distance
+            is considered infinite and not calculated.
+        samples : float, int
+            Number of samples (int) or fraction of coords to sample (float < 1).
+        """
+        
+        self.coords = coords.copy()
+        self.dist_metric = dist_metric
+        self.max_dist = max_dist
+        self.samples = samples
+
+        self._lidx = None
+        self._ridx = None
+        self._ltree = None
+        self._rtree = None
+        self._dists = None
+        # Do a very quick check to see throw exceptions if self.dist_metric is invalid...
+        pdist(self.coords[:1,:], metric=self.dist_metric)
+
+    @property
+    def sample_count(self):
+        if isinstance(self.samples, int):
+            return self.samples
+        return int(self.samples * len(self.coords))
+        
+    @property
+    def lidx(self):
+        """The sampled indices into `self.coords` for the left sample."""
+        if self._lidx is None:
+            self._lidx = np.random.choice(len(self.coords), size=self.sample_count, replace=False)
```

Comment:
Then, here the `self.seed` can be passed down.


BTW. numpy is suggesting to use [Generator](https://numpy.org/doc/stable/reference/random/generated/numpy.random.Generator.choice.html#numpy.random.Generator.choice) over `np.random.choice` today...